### DISCUSSION TITLE
The O3DE Editor does not start because of asserts and a nullptr exception

### DISCUSSION THREAD

**Discussion by maxmus1983** [o3de/o3de#16362] [Help Wanted] [answered]
Hello,

I have downloaded the source for O3DE and have build it sucessfully and I have created a sample project. In this sample project I wanted to launch the "Editor" as I assume that is the main Application to work with (aside from VS 2022). In this sample project everything builds fine in Visual Studio, but when I start the Editor I get an assert in 
RenderPipelinePtr RenderPipeline::CreateRenderPipeline(const RenderPipelineDescriptor& desc)
with AZ_Assert(pipeline->m_passTree.m_rootPass != nullptr, "Error creating root pass for pipeline!");

The Error log is:
Logged at Wed Jul 12 11:29:15 2023
FileVersion: 0.0.0.0
ProductVersion: 0.0.0.0
LogFile: @log@/Editor.log
ProjectDir: C:\Users\rantanplan2017\O3DE\Projects\AnimationTest
Executable: Editor.exe (debug: no
Exception Code: 0xC0000005
Exception Addr: 0x0033:0x00007FF8E827727E
Exception Module: 
Exception Description: EXCEPTION_ACCESS_VIOLATION, Attempt to read from address 0x00000000

The memory could not be "read"

Call Stack Trace:
18) 00007FF9C2690E4A (KERNELBASE) : UnhandledExceptionFilter
17) 00007FF9C56C58F2 (ntdll) : memset
16) 00007FF9C56AD376 (ntdll) : _C_specific_handler
15) 00007FF9C56C1FFF (ntdll) : _chkstk
14) 00007FF9C568AAD9 (ntdll) : RtlRaiseException
13) 00007FF9C56C0C6E (ntdll) : KiUserExceptionDispatcher
12) Z:\Libs\O3DE\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\RenderPipeline.cpp (175) : AZ::RPI::RenderPipeline::InitializeRenderPipeline
11) Z:\Libs\O3DE\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\RenderPipeline.cpp (55) : AZ::RPI::RenderPipeline::CreateRenderPipeline
10) Z:\Libs\O3DE\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\PreviewRenderer\PreviewRenderer.cpp (68) : AtomToolsFramework::PreviewRenderer::PreviewRenderer
 9) Z:\Libs\O3DE\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\PreviewRenderer\PreviewRendererSystemComponent.cpp (67) : `AtomToolsFramework::PreviewRendererSystemComponent::Activate'::`2'::<lambda_1>::operator()
 8) Z:\Libs\O3DE\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::TickEvents,AZ::TickEvents>,AZStd::recursive_mutex>::Execute
 7) Z:\Libs\O3DE\o3de\Code\Editor\CryEdit.cpp (1641) : CCryEditApp::InitInstance
 6) Z:\Libs\O3DE\o3de\Code\Editor\CryEdit.cpp (3822) : CryEditMain
 5) Z:\Libs\O3DE\o3de\Code\Editor\main.cpp (31) : main
 4) D:\github\3p-package-source\package-system\Qt\temp\src\qtbase\src\winmain\qtmain_win.cpp (97) : WinMain
 3) D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl (288) : __scrt_common_main_seh
 2) 00007FF9C3F57C24 (KERNEL32) : BaseThreadInitThunk
 1) 00007FF9C568D721 (ntdll) : RtlUserThreadStart

**Comment by AMZN-daimini** [comment]
@o3de/sig-graphics-audio

**Reply by maxmus1983** [reply]
I also got the following errors in the Editor.txt log

<19:26:22> [Error] (AssetUtils) - Could not compile asset 'Passes/PassTemplates.azasset', status = 1.
<19:26:22> [Error] (AssetUtils) - Unable to find product asset 'Passes/PassTemplates.azasset'. Has the source asset finished building?
<19:26:22> (RPI system) - System assets initialized

<19:26:22> [Error] (PassFactory) - FAILED TO CREATE PASS [PreviewRendererSystemComponent Preview Pipeline]. Could not find pass template [ToolsPipelineRenderToTexture]
<19:26:22> (System) -

**Comment by moudgils** [comment, marked as answer]
This is fixed by https://github.com/o3de/o3de/pull/16360

**Reply by maxmus1983** [reply]
Thanks for the quick answer!

Now I get another exception in
HRESULT MemoryBlock::Init(ID3D12ProtectedResourceSession* pProtectedSession, bool denyMsaaTextures)

hr = m_Allocator->GetDevice4()->CreateHeap1(&heapDesc, pProtectedSession, D3D12MA_IID_PPV_ARGS(&m_Heap));

throws off an unhandled exception. pProtectedSession is 0, maybe that is the reason.

The stack is

>	Atom_RHI_DX12.Private.dll!D3D12MA::MemoryBlock::Init(ID3D12ProtectedResourceSession * pProtectedSession, bool denyMsaaTextures) Line 8087	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::NormalBlock::Init(unsigned int algorithm, ID3D12ProtectedResourceSession * pProtectedSession, bool denyMsaaTextures) Line 8132	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::BlockVector::CreateBlock(unsigned __int64 blockSize, unsigned __int64 * pNewBlockIndex) Line 8795	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::BlockVector::AllocatePage(unsigned __int64 size, unsigned __int64 alignment, const D3D12MA::ALLOCATION_DESC & allocDesc, D3D12MA::Allocation * * pAllocation) Line 8686	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::BlockVector::Allocate(unsigned __int64 size, unsigned __int64 alignment, const D3D12MA::ALLOCATION_DESC & allocDesc, unsigned __int64 allocationCount, D3D12MA::Allocation * * pAllocations) Line 8312	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::BlockVector::CreateResource(unsigned __int64 size, unsigned __int64 alignment, const D3D12MA::ALLOCATION_DESC & allocDesc, const D3D12_RESOURCE_DESC & resourceDesc, D3D12_RESOURCE_STATES InitialResourceState, const D3D12_CLEAR_VALUE * pOptimizedClearValue, D3D12MA::Allocation * * ppAllocation, const _GUID & riidResource, void * * ppvResource) Line 8410	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::AllocatorPimpl::CreateResource(const D3D12MA::ALLOCATION_DESC * pAllocDesc, const D3D12_RESOURCE_DESC * pResourceDesc, D3D12_RESOURCE_STATES InitialResourceState, const D3D12_CLEAR_VALUE * pOptimizedClearValue, D3D12MA::Allocation * * ppAllocation, const _GUID & riidResource, void * * ppvResource) Line 6775	C++
 	Atom_RHI_DX12.Private.dll!D3D12MA::Allocator::CreateResource(const D3D12MA::ALLOCATION_DESC * pAllocDesc, const D3D12_RESOURCE_DESC * pResourceDesc, D3D12_RESOURCE_STATES InitialResourceState, const D3D12_CLEAR_VALUE * pOptimizedClearValue, D3D12MA::Allocation * * ppAllocation, const _GUID & riidResource, void * * ppvResource) Line 9953	C++
 	Atom_RHI_DX12.Private.dll!AZ::DX12::Device::CreateD3d12maBuffer(const AZ::RHI::BufferDescriptor & bufferDescriptor, D3D12_RESOURCE_STATES initialState, D3D12_HEAP_TYPE heapType) Line 573	C++
 	Atom_RHI_DX12.Private.dll!AZ::DX12::BufferD3D12MemoryAllocator::Allocate(unsigned __int64 sizeInBytes, unsigned __int64 overrideSubAllocAlignment) Line 54	C++
 	Atom_RHI_DX12.Private.dll!AZ::DX12::BufferPool::InitBufferInternal(AZ::RHI::Buffer & bufferBase, const AZ::RHI::BufferDescriptor & bufferDescriptor) Line 263	C++
 	Atom_RPI.Editor.dll!AZ::RHI::BufferPool::InitBuffer::__l2::<lambda_1>::operator()() Line 142	C++

**Reply by moudgils** [reply]
Seems like a crash within D3D12MA. Check the log for any warnings/asserts/output that can provide leads.

**Reply by maxmus1983** [reply]
The output with errors is:
D3D12 ERROR: ID3D12Device::CreateHeap1: D3D12_HEAP_FLAGS has recognized flags set. The value is 0x10c0, and the following flags are unrecognized: 0x1000. [ STATE_CREATION ERROR #630: CREATEHEAP_UNRECOGNIZEDMISCFLAGS]
D3D12: **BREAK** enabled for the previous message, which was: [ ERROR STATE_CREATION #630: CREATEHEAP_UNRECOGNIZEDMISCFLAGS ]
Exception thrown at 0x00007FF9C25D3B19 (KernelBase.dll) in Editor.exe: 0x0000087A (parameters: 0x0000000000000001, 0x000000A8EF336BB0, 0x000000A8EF3389A0).
Unhandled exception at 0x00007FF9C25D3B19 (KernelBase.dll) in Editor.exe: 0x0000087A (parameters: 0x0000000000000001, 0x000000A8EF336BB0, 0x000000A8EF3389A0).

This is quite strange as 0x1000 is
D3D12_HEAP_FLAG_CREATE_NOT_ZEROED	= 0x1000,
in enum D3D12_HEAP_FLAGS

**Reply by maxmus1983** [reply]
So in MemoryBlock::MemoryBlock(

I set 

m_HeapFlags(D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS),

This removes the flag that DX complains about (This flag seemed optional anyway). This fixes the problem and I can start the editor now.

Thanks for the quick help. I will mark this as fixed for now. I you want to find out why this happened so it does not also happen to others I will help you as I can.
